Don't use KUBE-MARK-DROP for "no local endpoints"

danwinship opened the pull request:
#### What type of PR is this?
/kind cleanup
/kind feature

#### What this PR does / why we need it:
As part of [KEP-3178](https://github.com/kubernetes/enhancements/issues/3178) this removes the use of `KUBE-MARK-DROP` for services with `Local` traffic policy and no local endpoints, by just having a `filter` table rule that calls `-j DROP`, exactly like the way we call `-j REJECT` when there are no endpoints at all.

I don't totally love how this came out... my original version did, eg

```
if has clusterIP endpoints {
    write clusterIP endpoints
} else if iTP:Local and has remote endpoints but not local endpoints {
    write clusterIP DROP rule
} else {
    write clusterIP REJECT rule
}
```

but the DROP/REJECT rules were always nearly identical so it seemed to make sense to try to refactor that a little, but... meh... I guess it does make it clear that the DROP and REJECT cases work in exactly the same way...

(All the tweaks to the `if` conditions in the patch are (a) to prevent us from creating the SVC/SVL chains if they would be empty now that there's no KUBE-MARK-DROP rule there, and (b) to prevent us from creating jumps to the SVC/SVL chains if we're not creating the SVC/SVL chains.)

#### Which issue(s) this PR fixes:
none

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3178
```

danwinship committed:
proxy/iptables: Don't use KUBE-MARK-DROP for "no local endpoints"

Rather than marking packets to be dropped in the "nat" table and then
dropping them from the "filter" table later, just use rules in
"filter" to drop the packets we don't like directly.

danwinship committed:
proxy/iptables: update TestTracePackets

Re-sync the rules from TestOverallIPTablesRulesWithMultipleServices to
make sure we're testing all the right kinds of rules. Remove a
duplicate copy of the KUBE-MARK-MASQ and KUBE-POSTROUTING rules.

Update the "REJECT" test to use the new svc6 from
TestOverallIPTablesRulesWithMultipleServices. (Previously it had used
a modified version of TOIPTRWMS's svc3.)

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1707,7 +1714,6 @@ func TestTracePackets(t *testing.T) {
 		-A KUBE-SVC-GNZBNJ2PO5MGZ6GT -m comment --comment "ns2/svc2:p80 -> 10.180.0.2:80" -j KUBE-SEP-RS4RBKLTHTF2IUXJ
 		-A KUBE-SVC-XPGD46QRK7WJZT7O -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 172.30.0.41 --dport 80 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
 		-A KUBE-SVC-XPGD46QRK7WJZT7O -m comment --comment "ns1/svc1:p80 -> 10.180.0.1:80" -j KUBE-SEP-SXIVWICOYRO3J4NJ
-		-A KUBE-SVL-GNZBNJ2PO5MGZ6GT -m comment --comment "ns2/svc2:p80 has no local endpoints" -j KUBE-MARK-DROP
```
L1695 		`-A KUBE-EXT-GNZBNJ2PO5MGZ6GT -j KUBE-SVL-GNZBNJ2PO5MGZ6GT`

danwinship committed:
proxy/iptables: fix up TestOverallIPTablesRulesWithMultipleServices

svc2b was using the same ClusterIP as svc3; change it and rename the
service to svc5 to make everything clearer.

Move the test of LoadBalancerSourceRanges from svc2 to svc5, so that
svc2 tests the rules for dropping packets due to
externalTrafficPolicy, and svc5 tests the rules for dropping packets
due to LoadBalancerSourceRanges, rather than having them both mixed
together in svc2.

Add svc6 with no endpoints.

thockin reviewed (commented):
Thanks!

/lgtm
/approve

wangycc commented:
Why does this PR drop the traffic accessing the LB VIP?
```golang
			// Either no endpoints at all (REJECT) or no endpoints for
			// external traffic (DROP anything that didn't get short-circuited
			// by the EXT chain.)
			for _, lbip := range svcInfo.LoadBalancerIPStrings() {
				proxier.filterRules.Write(
					"-A", string(kubeExternalServicesChain),
					"-m", "comment", "--comment", fmt.Sprintf(`"%s has no endpoints"`, svcPortNameString),
					"-m", "comment", "--comment", externalTrafficFilterComment,
					"-m", protocol, "-p", protocol,
					"-d", lbip,
					"--dport", strconv.Itoa(svcInfo.Port()),
					"-j", "REJECT",
					"-j", externalTrafficFilterTarget,
				)
			}
		}
```

danwinship commented:
There should not have been any user-visible changes in behavior. If you think there was, please file a bug with the details.
